### PR TITLE
Require explicit OPTIONAL qualifiers for properties and links

### PR DESCRIPTION
**Pull request by ambv**
I also mass transformed unqualified properties and links in tests. OPTIONAL is now also included in the generated DDL. Documentation updated.

### CODE REVIEW

File: docs/edgeql/ddl/links.rst [hunk 1]
```diff
@@ -216,7 +223,7 @@ Change the definition of a :ref:`link <ref_datamodel_links>`.
       CREATE ANNOTATION <annotation-name> := <value>
       ALTER ANNOTATION <annotation-name> := <value>
       DROP ANNOTATION <annotation-name>
-      CREATE PROPERTY <property-name> ...
+      CREATE OPTIONAL PROPERTY <property-name> ...
```
Comment (1st1) [inline docs/edgeql/ddl/links.rst:226]:
```suggestion
      CREATE {REQUIRED|OPTIONAL}  PROPERTY <property-name> ...
```

File: docs/edgeql/ddl/links.rst [hunk 2]
```diff
@@ -272,7 +279,7 @@ The following subcommands are allowed in the ``ALTER LINK`` block:
     Make the link *required*.
 
 :eql:synopsis:`DROP REQUIRED`
-    Make the link no longer *required*.
+    Make the link *optional*, i.e. no longer *required*.
```
Comment (1st1) [inline docs/edgeql/ddl/links.rst:282]: We have to fix this in a follow up PR. @elprans

File: docs/edgeql/ddl/objects.rst [hunk 3]
```diff
@@ -133,10 +133,10 @@ Change the definition of an
       CREATE ANNOTATION <annotation-name> := <value>
       ALTER ANNOTATION <annotation-name> := <value>
       DROP ANNOTATION <annotation-name>
-      CREATE LINK <link-name> ...
+      CREATE OPTIONAL LINK <link-name> ...
```
Comment (1st1) [inline docs/edgeql/ddl/objects.rst:136]:
```suggestion
      CREATE {REQUIRED|OPTIONAL} LINK <link-name> ...
```

File: docs/edgeql/ddl/objects.rst [hunk 4]
```diff
@@ -133,10 +133,10 @@ Change the definition of an
...
       ALTER LINK <link-name> ...
       DROP LINK <link-name> ...
-      CREATE PROPERTY <property-name> ...
+      CREATE OPTIONAL PROPERTY <property-name> ...
```
Comment (1st1) [inline docs/edgeql/ddl/objects.rst:139]:
```suggestion
      CREATE {REQUIRED|OPTIONAL} PROPERTY <property-name> ...
```

File: edb/edgeql/parser/grammar/sdl.py [hunk 5]
```diff
@@ -624,6 +624,7 @@ def reduce_CreateRegularProperty(self, *kids):
             target=kids[3].val,
             commands=kids[4].val,
         )
+        # reject_unqualified(self.val)
```
Comment (1st1) [inline edb/edgeql/parser/grammar/sdl.py:627]: ?

Should this be uncommented?

File: edb/schema/pointers.py [hunk 6]
```diff
@@ -1049,7 +1050,14 @@ def _process_create_or_alter_ast(
 
         This may be called in the context of either Create or Alter.
         """
-        if astnode.is_required is not None:
+        if astnode.is_required is None:
+            if not context.dumprestore:
+                raise EdgeQLSyntaxError(
+                    f'Provide an explicit REQUIRED or OPTIONAL qualifier for'
```
Comment (1st1) [inline edb/schema/pointers.py:1056]:
```suggestion
                    f'specify an explicit REQUIRED or OPTIONAL qualifier for'
```

File: edb/edgeql/parser/grammar/sdl.py (hunk 5 above)
Comment (ambv) [reply edb/edgeql/parser/grammar/sdl.py:627]: Yes, that generates a few more test failures that I'm working on now. I commented those for now so we can focus on the DROP REQUIRED test failures on CI.

File: docs/edgeql/ddl/links.rst (hunk 2 above)
Comment (ambv) [reply docs/edgeql/ddl/links.rst:282]: Yeah, we talked about it a few hours back.

Comment (1st1) [reply docs/edgeql/ddl/links.rst:282]: I do remember that. I'm reinforcing the fact that we need to fix this before a3.

**Comment by 1st1** [comment]
Closing this in favor of https://github.com/edgedb/edgedb/pull/1447. The new PR is based on @ambv's work with his commit being amended.

**Pull request closed without merging**